### DISCUSSION TITLE
Valhalla Kubernetes operator

### DISCUSSION THREAD

**Discussion by itayankri** [valhalla/valhalla#3802] [Show and tell]
Hi guys,

I wanted to share with the community a new tool I developed called valhalla-operator.
[itayankri/valhalla-operator](https://github.com/itayankri/valhalla-operator)

I think that I am not the only one that struggling with deploying valhalla on Kubernetes at scale, so I developed a Kubernetes operator that automates the process. It will help you scale up your Valhalla deployments, or even migrate them to Kubernetes if you like.

I also wrote a medium article that demonstrates how to use it.
[medium.com/@itay.ankri/deploying-valhalla-routing-engine-on-kubernetes-using-valhalla-operator-2426e79ac746](https://medium.com/@itay.ankri/deploying-valhalla-routing-engine-on-kubernetes-using-valhalla-operator-2426e79ac746)

Feel free to approach me with any feedback or suggestions you have.

Enjoy!

**Comment by kevinkreiser** [comment]
thanks for sharing this i hope others find it useful. its not documented but i can tell you a bit about the philosophy of the docker image we create and how its intended to be used in production.

the image has all of the tools valhalla contains to do the various things it can do. that means it has the tools to make tile data, update that data with other data, analytical tools to run routes etc in an offline manner and the tools to run the services in a fault tolerant way. in my view it didnt make sense for us to produce an image which prescribes the use-case for the user. instead we give you the tools and you use them for your use-case however you want.

regarding data creation and running the service with that data though id like to say that we view those as two separate processes. yes the first time you check out valhalla you're going to make data and then try routing on it, but in a production environment you should probably be making data every day in some offline processing and simply serving routes off of that data in some online process. what i mean is that those two processes are asynchronous and not exactly coordinated. when your k8s cron job that makes new routing data daily pushes that up to some storage in the cloud, you need to configure something to listen for that data to land and trigger a k8s deployment of the running service to pick up the latest data. this is what i have done in production though many other permutations could exist.

again, i know none of this is documented, but its kind of by design. we dont want to prescribe to anyone how to run their service or deploy their application. we would like the tools we build to be generic enough that people can string them together in a way that makes sense for their application architecture or cloud platform or whatever.
